controller/src/main/java/io/pravega/controller/store/stream/PersistentStreamBase.java

```diff
@@ -626,9 +627,22 @@ private long findSegmentSplitsMerges(List<StreamSegmentRecord> referenceSegments
 
     @Override
     public CompletableFuture<Boolean> isStreamCutValid(Map<Long, Long> streamCut) {
-        return Futures.allOfWithResults(streamCut.keySet().stream().map(x -> getSegment(x).thenApply(segment ->
-                new SimpleEntry<>(segment.getKeyStart(), segment.getKeyEnd())))
-                                                 .collect(Collectors.toList()))
+        Map<Integer, List<Long>> groupByEpoch = streamCut.keySet().stream().collect(groupingBy(StreamSegmentNameUtils::getEpoch));
```

Comment:
According to the PR description, this change is avoiding many `getEpochRecord` calls to the cache. Could you elaborate? I don't see it.

Reply:
we dont store segmentRecords at the top level but within an epochRecord. 
EpochRecords are immutable and cached. 
Here, the original implementation called getEpochRecord for all segments. So if there were multiple segments from the same epoch, the call will be made multiple times. 

by first grouping segments by their epochs, and then making a single call per epoch, we improve the efficiency there.